src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -115,105 +113,147 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
+                // what actually changed.
+                using var _6 = PooledHashSet<Checksum>.GetInstance(out var newChecksumsToSync);
+                newChecksumsToSync.AddRange(newProjectIdToChecksum.Values);
 
-                    Contract.ThrowIfTrue(oldProjectChecksums.Checksum == newProjectChecksums.Checksum);
+                var newProjectStateChecksums = await _assetProvider.GetAssetsAsync<ProjectStateChecksums>(
+                    assetHint: AssetHint.None, newChecksumsToSync, cancellationToken).ConfigureAwait(false);
 
-                    solution = await UpdateProjectAsync(solution.GetRequiredProject(projectId), oldProjectChecksums, newProjectChecksums, cancellationToken).ConfigureAwait(false);
+                foreach (var (checksum, newProjectStateChecksum) in newProjectStateChecksums)
+                {
+                    Contract.ThrowIfTrue(checksum != newProjectStateChecksum.Checksum);
+                    var projectId = newProjectStateChecksum.ProjectId;
+                    newProjectIdToStateChecksums.Add(projectId, newProjectStateChecksum);
                 }
 
+                // Now that we've collected the old and new project state checksums, we can actually process them to
+                // determine what to remove, what to add, and what to change.
+                solution = await UpdateProjectsAsync(
+                    solution, isConeSync, oldProjectIdToStateChecksums, newProjectIdToStateChecksums, cancellationToken).ConfigureAwait(false);
+
                 return solution;
+            }
 
-                async Task PopulateOldProjectMapAsync()
+            private async Task<Solution> UpdateProjectsAsync(
+                Solution solution,
+                bool isConeSync,
+                Dictionary<ProjectId, ProjectStateChecksums> oldProjectIdToStateChecksums,
+                Dictionary<ProjectId, ProjectStateChecksums> newProjectIdToStateChecksums,
+                CancellationToken cancellationToken)
+            {
+                // added project
+                foreach (var (projectId, newProjectChecksums) in newProjectIdToStateChecksums)
                 {
-                    foreach (var (projectId, projectState) in solution.SolutionState.ProjectStates)
+                    if (!oldProjectIdToStateChecksums.ContainsKey(projectId))
                     {
-                        var projectChecksums = await projectState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-                        if (olds.Object.Contains(projectChecksums.Checksum))
-                            oldMap.Add(projectId, projectChecksums);
+                        // bulk sync added project assets fully since we'll definitely need that data, and we won't want
+                        // to make tons of intermediary calls for it.
+
+                        await _assetProvider.SynchronizeProjectAssetsAsync(newProjectChecksums, cancellationToken).ConfigureAwait(false);
```

Comment:
note: this was a step i errantly removed when doing the original change here.  So when we saw a project was being added, we'd end up making *tons* of syncing calls for it for all its children, instead of doing it in single batch calls per projet.

Reply:
i also moved this here for clarity.  the original code did all this first.  note: it still walked a single project at a time syncing all its assets.  we could consider doing all that in one step in bulk as well (and i'll likely do that as a followup).